In [166]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import matplotlib as mpl

In [167]:
filepath = '/home/daria/REALEC/realec_72.csv'

In [168]:
data = pd.read_csv(filepath, delimiter=';')

## Preprocessing

In [169]:
data['_name'] = data['name'].apply(lambda r: r.split('_')[-1])

In [187]:
groups = data.groupby('_name')
frst = groups.get_group('1.txt').drop(['_name'], axis=1)
scnd = groups.get_group('2.txt').drop(['_name'], axis=1)

In [171]:
frst

name   CEFR  av_depth  max_depth  min_depth  num_acl  \
0      AAl_1_1.txt  B1-B2  4.272727          7          2        7   
3     AAl_12_1.txt     B1  4.923077          8          1        4   
5     AAl_14_1.txt     B2  4.500000         10          3        4   
6     AAl_15_1.txt     B1  5.888889          9          5        4   
8     AAl_16_1.txt  B1-B2  4.636364          8          2        3   
...            ...    ...       ...        ...        ...      ...   
1162  VPe_32_1.txt     B1  4.285714          5          3        0   
1164  VPe_33_1.txt  A2-B1  7.600000         11          5        5   
1167  VPe_35_1.txt     B1  3.714286          7          2        1   
1169  VPe_36_1.txt  A2-B1  4.000000          6          3        4   
1171  KKo_24_1.txt     B2  6.000000          9          4        2   

      num_rel_cl  num_advcl  num_sent  num_tok  ...  num_4grams  \
0              1          5        11      220  ...           6   
3              1          7        13      314  ...           5   
5              1          0        10      229  ...           6   
6              0          3         9      252  ...           0   
8              1          2        11      246  ...          10   
...          ...        ...       ...      ...  ...         ...   
1162           0          1         7      161  ...           0   
1164           1          2         5      174  ...           3   
1167           5          1        14      240  ...           3   
1169           2          3         7      188  ...           1   
1171           1          6         6      228  ...           3   

      num_func_ngrams  num_shell_noun  num_misspelled_tokens  \
0                   1               2                     13   
3                   0               0                     12   
5                   0               0                      2   
6                   0               0                      5   
8                   3               0                      4   
...               ...             ...                    ...   
1162                0               0                      2   
1164                0               0                      7   
1167                0               0                      3   
1169                0               0                      4   
1171                1               0                      2   

      punct_mistakes_pp  punct_mistakes_because  punct_mistakes_but  \
0                     0                       0                   1   
3                     0                       0                   0   
5                     0                       0                   0   
6                     1                       0                   0   
8                     0                       0                   1   
...                 ...                     ...                 ...   
1162                  0                       0                   0   
1164                  0                       0                   0   
1167                  0                       0                   0   
1169                  0                       1                   3   
1171                  2                       0                   1   

      punct_mistakes_compare  million_mistake  side_mistake  
0                          0                0             0  
3                          0                0             0  
5                          0                0             0  
6                          0                0             0  
8                          0                0             0  
...                      ...              ...           ...  
1162                       0                0             0  
1164                       0                0             0  
1167                       0                0             0  
1169                       0                0             0  
1171                       0                1             0  

[

In [188]:
scnd

name   CEFR  av_depth  max_depth  min_depth  num_acl  \
1      AAl_1_2.txt     B1  4.166667          7          0        1   
2     AAl_11_2.txt     B1  4.142857          8          1        7   
4     AAl_12_2.txt     B1  4.555556          9          0        4   
7     AAl_15_2.txt     B1  4.416667         10          0        4   
9     AAl_16_2.txt  B1-B2  3.642857          8          0        4   
...            ...    ...       ...        ...        ...      ...   
1163  VPe_32_2.txt  B1-B2  4.400000          7          3        3   
1165  VPe_33_2.txt     B1  5.000000         12          0        8   
1166  VPe_34_2.txt     B1  4.529412          8          1        4   
1168  VPe_35_2.txt     B1  4.000000          8          0        6   
1170  AAl_14_2.txt  B1-B2  5.461538         10          0        6   

      num_rel_cl  num_advcl  num_sent  num_tok  ...  num_4grams  \
1              0          6        18      336  ...           4   
2              0          6        21      359  ...          11   
4              2          6        18      354  ...          13   
7              3          2        12      291  ...           3   
9              1          6        14      315  ...           6   
...          ...        ...       ...      ...  ...         ...   
1163           2          4        10      221  ...           8   
1165           0         10        12      339  ...           6   
1166           2          9        17      368  ...           4   
1168           1          5        18      387  ...          11   
1170           3          9        13      341  ...          13   

      num_func_ngrams  num_shell_noun  num_misspelled_tokens  \
1                   1               4                     12   
2                   2               5                      8   
4                   1               1                     15   
7                   0               1                      7   
9                   0               7                      4   
...               ...             ...                    ...   
1163                2               2                      3   
1165                3               1                     12   
1166                2               3                     15   
1168                2               5                      5   
1170                3               3                      7   

      punct_mistakes_pp  punct_mistakes_because  punct_mistakes_but  \
1                     0                       0                   4   
2                     0                       0                   2   
4                     1                       1                   1   
7                     0                       0                   0   
9                     0                       3                   0   
...                 ...                     ...                 ...   
1163                  0                       0                   0   
1165                  1                       0                   1   
1166                  1                       1                   0   
1168                  0                       0                   0   
1170                  0                       0                   0   

      punct_mistakes_compare  million_mistake  side_mistake  
1                          0                0             0  
2                          1                0             0  
4                          0                0             0  
7                          0                0             0  
9                          0                0             0  
...                      ...              ...           ...  
1163                       0                0             0  
1165                       0                0             0  
1166                       2                0             0  
1168                       0                0             0  
1170                       0                0             0  

[

## Multi Layer Perceptron Classifier

In [139]:
scaler = StandardScaler()
fof = scaler.fit_transform(frst.iloc[:, 2:])
X_train, X_test, y_train, y_test = train_test_split(fof, frst['CEFR'], test_size=0.25)

In [146]:
import matplotlib.pyplot as plt
from sklearn.datasets import fetch_openml
from sklearn.exceptions import ConvergenceWarning
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split



mlp = MLPClassifier(
    hidden_layer_sizes=(1000,)
    ,activation='relu',
    solver='adam',
    alpha=0.1, batch_size='auto',
    learning_rate='constant', learning_rate_init=0.004,
    power_t=0.5, max_iter=200, shuffle=True, random_state=None,
    tol=0.0001, verbose=False, warm_start=True, momentum=0.9, 
    nesterovs_momentum=True, early_stopping=False,
    validation_fraction=0.1, beta_1=0.9, beta_2=0.999, 
    epsilon=1e-08, n_iter_no_change=10, max_fun=15000
)
mlp.fit(X_train, y_train)

print("Training set score: %f" % mlp.score(X_train, y_train))
print("Test set score: %f" % mlp.score(X_test, y_test))

Training set score: 0.997685
Test set score: 0.263889


## Extra Trees Classifier

In [151]:
from sklearn.model_selection import cross_val_score
from sklearn.datasets import make_blobs
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.tree import DecisionTreeClassifier


clf = ExtraTreesClassifier(n_estimators=1000, max_depth=None,
    min_samples_split=2, random_state=0)
scores = cross_val_score(clf, fof, frst['CEFR'], cv=5)




/home/daria/.local/lib/python3.9/site-packages/sklearn/model_selection/_split.py:676: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=5.
  warnings.warn(


In [153]:
scores.mean()

0.37322338830584706

## Histogram-based Gradient Boosting Classification Tree

In [155]:
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.ensemble import VotingClassifier

clf = HistGradientBoostingClassifier(max_iter=100)
scores = cross_val_score(clf, fof, frst['CEFR'], cv=5)
scores.mean()

/home/daria/.local/lib/python3.9/site-packages/sklearn/model_selection/_split.py:676: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=5.
  warnings.warn(


0.3229535232383808

In [158]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import VotingClassifier

clf1 = LogisticRegression(random_state=1)
clf2 = RandomForestClassifier(n_estimators=50, random_state=1)
clf3 = GaussianNB()

eclf = VotingClassifier(
    estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)],
    voting='soft')

scores = cross_val_score(eclf, fof, frst['CEFR'], cv=5)
scores.mean()

/home/daria/.local/lib/python3.9/site-packages/sklearn/model_selection/_split.py:676: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=5.
  warnings.warn(
/home/daria/.local/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/daria/.local/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-

0.2674812593703148

## K Neighbors Classifier

In [165]:

from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=40)
scores = cross_val_score(neigh, fof, frst['CEFR'], cv=100)
scores.mean()


/home/daria/.local/lib/python3.9/site-packages/sklearn/model_selection/_split.py:676: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=100.
  warnings.warn(


0.38000000000000006

## CatBoost Classification

In [172]:
import catboost
from catboost import CatBoostClassifier

In [173]:
clf = CatBoostClassifier()

In [174]:
def classify(df):
    X_train, X_test, y_train, y_test = train_test_split(df.iloc[:, 2:], df['CEFR'], test_size=0.25)
    clf.fit(X_train, y_train, eval_set=(X_test, y_test))
    return clf.score(X_test, y_test)

## Classifying (second task)

In [176]:
classify(scnd)

Custom logger is already specified. Specify more than one logger at same time is not thread safe.

Learning rate set to 0.106251
0:	learn: 1.8865450	test: 1.9047428	best: 1.9047428 (0)	total: 17.7ms	remaining: 17.7s
1:	learn: 1.8394947	test: 1.8724965	best: 1.8724965 (1)	total: 55ms	remaining: 27.4s
2:	learn: 1.7843625	test: 1.8325825	best: 1.8325825 (2)	total: 76.1ms	remaining: 25.3s
3:	learn: 1.7435162	test: 1.8120297	best: 1.8120297 (3)	total: 99.4ms	remaining: 24.8s
4:	learn: 1.6954736	test: 1.7718135	best: 1.7718135 (4)	total: 137ms	remaining: 27.3s
5:	learn: 1.6569554	test: 1.7471137	best: 1.7471137 (5)	total: 183ms	remaining: 30.4s
6:	learn: 1.6194507	test: 1.7163062	best: 1.7163062 (6)	total: 225ms	remaining: 31.9s
7:	learn: 1.5833563	test: 1.6975317	best: 1.6975317 (7)	total: 256ms	remaining: 31.7s
8:	learn: 1.5513342	test: 1.6739428	best: 1.6739428 (8)	total: 281ms	remaining: 30.9s
9:	learn: 1.5212194	test: 1.6615567	best: 1.6615567 (9)	total: 327ms	remaining: 32.4s
10:	learn: 1.4877053	test: 1.6383493	best: 1.6383493 (10)	total: 375ms	remaining: 33.7s
11:	learn: 1.4637528

0.40268456375838924

In [177]:
feat_imp_scnd = clf.get_feature_importance(prettified=True)
feat_imp_scnd

Feature Id  Importances
0    num_misspelled_tokens     6.636251
1               der_level3     3.840139
2                       ls     3.624146
3       av_tok_before_root     3.081543
4                      ndw     2.971151
..                     ...          ...
67                  num_cl     0.210220
68           num_pres_plur     0.066405
69  punct_mistakes_compare     0.000000
70         million_mistake     0.000000
71            side_mistake     0.000000

[72 rows x 2 columns]

## Identifying and removing weak features (importance of less than 1)

In [178]:
weak_features = feat_imp_scnd[feat_imp_scnd['Importances'] < 1]
weak_features

Feature Id  Importances
43             pos_sim_nei     0.979543
44                       d     0.940752
45           num_part_noun     0.902451
46                 num_inf     0.874397
47            num_adj_noun     0.842609
48              num_rel_cl     0.808552
49       punct_mistakes_pp     0.808492
50           lemma_sim_nei     0.787370
51                  num_tu     0.778370
52            corrected_vv     0.760501
53  punct_mistakes_because     0.729523
54                 log_ttr     0.721021
55                 num_acl     0.715818
56               max_depth     0.703887
57                lfp_1000     0.635100
58       freq_finite_forms     0.594864
59         num_func_ngrams     0.565079
60           corrected_ttr     0.555107
61                num_poss     0.458576
62                num_sent     0.435504
63             pos_sim_all     0.409754
64            corrected_vs     0.409591
65               min_depth     0.269953
66              der_level6     0.255005
67                  num_cl     0.210220
68           num_pres_plur     0.066405
69  punct_mistakes_compare     0.000000
70         million_mistake     0.000000
71            side_mistake     0.000000

In [179]:
scnd = scnd.drop(weak_features['Feature Id'], axis=1)
scnd

name   CEFR  av_depth  num_advcl  num_tok  av_tok_before_root  \
1      AAl_1_2.txt     B1  4.166667          6      336            5.111111   
2     AAl_11_2.txt     B1  4.142857          6      359            5.047619   
4     AAl_12_2.txt     B1  4.555556          6      354            5.277778   
7     AAl_15_2.txt     B1  4.416667          2      291            8.250000   
9     AAl_16_2.txt  B1-B2  3.642857          6      315            5.357143   
...            ...    ...       ...        ...      ...                 ...   
1163  VPe_32_2.txt  B1-B2  4.400000          4      221            5.400000   
1165  VPe_33_2.txt     B1  5.000000         10      339            7.833333   
1166  VPe_34_2.txt     B1  4.529412          9      368            6.117647   
1168  VPe_35_2.txt     B1  4.000000          5      387            5.722222   
1170  AAl_14_2.txt  B1-B2  5.461538          9      341            5.384615   

      av_len_sent  num_compl_tu  num_coord  num_prep  ...  freq_aux  \
1       18.666667             7         11        12  ...  0.781250   
2       17.095238            10         15        25  ...  0.513514   
4       19.666667             8          3        21  ...  0.764706   
7       24.250000             5          9        18  ...  0.869565   
9       22.500000            10          7         9  ...  0.862069   
...           ...           ...        ...       ...  ...       ...   
1163    22.100000             6         10        13  ...  0.307692   
1165    28.250000            13          8         7  ...  0.696970   
1166    21.647059            11          7        21  ...  0.771429   
1168    21.500000             9         21        20  ...  0.678571   
1170    26.230769            10          7        21  ...  0.806452   

      num_gerunds  num_pres_sing  num_past_part  num_past_simple  \
1               1              9              8                1   
2               5             11              4                8   
4               3             12              5                4   
7               0              5              6                4   
9               1              6              1                9   
...           ...            ...            ...              ...   
1163            3              8              1                5   
1165            1              6              5                5   
1166            6              9              3               10   
1168            1              2              1                6   
1170            1             12              6                5   

      num_linkings  num_4grams  num_shell_noun  num_misspelled_tokens  \
1               13           4               4                     12   
2               15          11               5                      8   
4               15          13               1                     15   
7               12           3               1                      7   
9               16           6               7                      4   
...            ...         ...             ...                    ...   
1163            13           8               2                      3   
1165            23           6               1                     12   
1166            25           4               3                     15   
1168            22          11               5                      5   
1170            25          13               3                      7   

      punct_mistakes_but  
1                      4  
2                      2  
4                      1  
7                      0  
9                      0  
...                  ...  
1163                   0  
1165                   1  
1166                   0  
1168                   0  
1170                   0  

[596 rows x 45 columns]

In [181]:
classify(scnd)

Learning rate set to 0.106251
0:	learn: 1.8907423	test: 1.9065557	best: 1.9065557 (0)	total: 29.2ms	remaining: 29.1s
1:	learn: 1.8368807	test: 1.8574363	best: 1.8574363 (1)	total: 60.9ms	remaining: 30.4s
2:	learn: 1.7770947	test: 1.8029982	best: 1.8029982 (2)	total: 100ms	remaining: 33.3s
3:	learn: 1.7303721	test: 1.7602807	best: 1.7602807 (3)	total: 133ms	remaining: 33.1s
4:	learn: 1.6948888	test: 1.7402746	best: 1.7402746 (4)	total: 167ms	remaining: 33.2s
5:	learn: 1.6575325	test: 1.7125627	best: 1.7125627 (5)	total: 191ms	remaining: 31.6s
6:	learn: 1.6192504	test: 1.6813369	best: 1.6813369 (6)	total: 225ms	remaining: 31.9s
7:	learn: 1.5828070	test: 1.6548523	best: 1.6548523 (7)	total: 243ms	remaining: 30.1s
8:	learn: 1.5533333	test: 1.6344439	best: 1.6344439 (8)	total: 261ms	remaining: 28.7s
9:	learn: 1.5229284	test: 1.6125337	best: 1.6125337 (9)	total: 288ms	remaining: 28.5s
10:	learn: 1.5010990	test: 1.6013252	best: 1.6013252 (10)	total: 312ms	remaining: 28s
11:	learn: 1.4750379	t

0.42953020134228187

## Second task:

**Classifier score BEFORE removing weak features is 0.40268456375838924**

**Classifier score AFTER removing weak features is 0.42953020134228187**

## Classifying (first task)

In [182]:
classify(frst)

Learning rate set to 0.106179
0:	learn: 1.7385811	test: 1.7392708	best: 1.7392708 (0)	total: 14.5ms	remaining: 14.5s
1:	learn: 1.7001818	test: 1.7113456	best: 1.7113456 (1)	total: 38.8ms	remaining: 19.4s
2:	learn: 1.6595533	test: 1.6757688	best: 1.6757688 (2)	total: 62.7ms	remaining: 20.8s
3:	learn: 1.6241927	test: 1.6437486	best: 1.6437486 (3)	total: 78.2ms	remaining: 19.5s
4:	learn: 1.5935999	test: 1.6202755	best: 1.6202755 (4)	total: 92.9ms	remaining: 18.5s
5:	learn: 1.5624135	test: 1.6022805	best: 1.6022805 (5)	total: 109ms	remaining: 18s
6:	learn: 1.5343821	test: 1.5838132	best: 1.5838132 (6)	total: 137ms	remaining: 19.4s
7:	learn: 1.5084700	test: 1.5711458	best: 1.5711458 (7)	total: 155ms	remaining: 19.2s
8:	learn: 1.4842287	test: 1.5583564	best: 1.5583564 (8)	total: 172ms	remaining: 19s
9:	learn: 1.4666391	test: 1.5416391	best: 1.5416391 (9)	total: 199ms	remaining: 19.7s
10:	learn: 1.4474376	test: 1.5273973	best: 1.5273973 (10)	total: 228ms	remaining: 20.5s
11:	learn: 1.4286487	

0.4375

In [183]:
feat_imp_frst = clf.get_feature_importance(prettified=True)
feat_imp_frst

Feature Id  Importances
0                 freq_aux     3.332017
1    num_misspelled_tokens     3.017284
2                       ls     2.955171
3            num_pres_sing     2.838389
4               der_level5     2.765959
..                     ...          ...
67                 num_acl     0.173995
68  punct_mistakes_because     0.058702
69  punct_mistakes_compare     0.041677
70           num_pres_plur     0.000000
71            side_mistake     0.000000

[72 rows x 2 columns]

## Identifying and removing weak features (importance of less than 0.5)

In [184]:
weak_features = feat_imp_frst[feat_imp_frst['Importances'] < 0.5]
weak_features

Feature Id  Importances
61               min_depth     0.458981
62       punct_mistakes_pp     0.419903
63         million_mistake     0.374136
64            num_compl_tu     0.322790
65          num_shell_noun     0.293576
66            num_adj_noun     0.232186
67                 num_acl     0.173995
68  punct_mistakes_because     0.058702
69  punct_mistakes_compare     0.041677
70           num_pres_plur     0.000000
71            side_mistake     0.000000

In [185]:
frst = frst.drop(weak_features['Feature Id'], axis=1)
frst

name   CEFR  av_depth  max_depth  num_rel_cl  num_advcl  \
0      AAl_1_1.txt  B1-B2  4.272727          7           1          5   
3     AAl_12_1.txt     B1  4.923077          8           1          7   
5     AAl_14_1.txt     B2  4.500000         10           1          0   
6     AAl_15_1.txt     B1  5.888889          9           0          3   
8     AAl_16_1.txt  B1-B2  4.636364          8           1          2   
...            ...    ...       ...        ...         ...        ...   
1162  VPe_32_1.txt     B1  4.285714          5           0          1   
1164  VPe_33_1.txt  A2-B1  7.600000         11           1          2   
1167  VPe_35_1.txt     B1  3.714286          7           5          1   
1169  VPe_36_1.txt  A2-B1  4.000000          6           2          3   
1171  KKo_24_1.txt     B2  6.000000          9           1          6   

      num_sent  num_tok  av_tok_before_root  av_len_sent  ...  num_inf  \
0           11      220            6.272727    20.000000  ...        2   
3           13      314            8.846154    24.153846  ...       11   
5           10      229            5.200000    22.900000  ...        2   
6            9      252            7.111111    28.000000  ...        1   
8           11      246            4.636364    22.363636  ...        8   
...        ...      ...                 ...          ...  ...      ...   
1162         7      161            6.571429    23.000000  ...        4   
1164         5      174            1.600000    34.800000  ...        3   
1167        14      240            5.500000    17.142857  ...        4   
1169         7      188           11.142857    26.857143  ...        4   
1171         6      228            3.833333    38.000000  ...        6   

      num_gerunds  num_pres_sing  num_past_part  num_past_simple  \
0               2              9              9                2   
3               3              8             12                5   
5               0              1              6               10   
6               3              4              7                5   
8               0             15              2                3   
...           ...            ...            ...              ...   
1162            0              7              0                3   
1164            0              4              5                2   
1167            3              1              5                9   
1169            2              8              8                2   
1171            3              1              4                2   

      num_linkings  num_4grams  num_func_ngrams  num_misspelled_tokens  \
0                6           6                1                     13   
3               13           5                0                     12   
5                8           6                0                      2   
6               14           0                0                      5   
8                5          10                3                      4   
...            ...         ...              ...                    ...   
1162             4           0                0                      2   
1164             9           3                0                      7   
1167            13           3                0                      3   
1169            10           1                0                      4   
1171            17           3                1                      2   

      punct_mistakes_but  
0                      1  
3                      0  
5                      0  
6                      0  
8                      1  
...                  ...  
1162                   0  
1164                   0  
1167                   0  
1169                   3  
1171                   1  

[576 rows x 63 columns]

In [186]:
classify(frst)

Learning rate set to 0.106179
0:	learn: 1.7439864	test: 1.7472619	best: 1.7472619 (0)	total: 11.9ms	remaining: 11.9s
1:	learn: 1.7025357	test: 1.7169173	best: 1.7169173 (1)	total: 26.3ms	remaining: 13.1s
2:	learn: 1.6700128	test: 1.6987468	best: 1.6987468 (2)	total: 43.7ms	remaining: 14.5s
3:	learn: 1.6362985	test: 1.6691683	best: 1.6691683 (3)	total: 68.6ms	remaining: 17.1s
4:	learn: 1.6064892	test: 1.6411405	best: 1.6411405 (4)	total: 95.5ms	remaining: 19s
5:	learn: 1.5795951	test: 1.6244491	best: 1.6244491 (5)	total: 123ms	remaining: 20.5s
6:	learn: 1.5495587	test: 1.6045814	best: 1.6045814 (6)	total: 153ms	remaining: 21.8s
7:	learn: 1.5227698	test: 1.5875305	best: 1.5875305 (7)	total: 183ms	remaining: 22.7s
8:	learn: 1.4980868	test: 1.5733470	best: 1.5733470 (8)	total: 208ms	remaining: 22.9s
9:	learn: 1.4776952	test: 1.5590675	best: 1.5590675 (9)	total: 224ms	remaining: 22.1s
10:	learn: 1.4566323	test: 1.5475913	best: 1.5475913 (10)	total: 247ms	remaining: 22.2s
11:	learn: 1.435467

0.4722222222222222

## First task:

**Classifier score BEFORE removing weak features is 0.4375**

**Classifier score BEFORE removing weak features is 0.4722222222222222**

## Bonus: optuna on CatBoost Classifier parameters

In [26]:
import optuna

def objective(trial):
    params = {
        'depth': trial.suggest_int("depth", 1, 12),
        'learning_rate': trial.suggest_float("learning_rate", 0.01, 1), 
        'l2_leaf_reg': trial.suggest_int("l2_leaf_reg", 1, 100000),
        'border_count': trial.suggest_int("border_count", 1, 10000),
        'random_strength': trial.suggest_int("random_strength", 1, 10000),
        'grow_policy': trial.suggest_categorical("grow_policy", ['SymmetricTree', 'Depthwise', 'Lossguide']),
        'leaf_estimation_iterations': trial.suggest_int("leaf_estimation_iterations", 1, 100),
        'leaf_estimation_method': trial.suggest_categorical("leaf_estimation_method", ['Newton', 'Gradient']),
        'bootstrap_type': trial.suggest_categorical("bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]),
        'score_function': trial.suggest_categorical("score_function", ["Cosine", "L2"]),
        'allow_const_label':True,
        'use_best_model': True,
        'od_type': 'IncToDec',
        }
    if params['grow_policy'] != 'SymmetricTree':
        params['boosting_type']: 'Plain'
    else:
        params['boosting_type']: 'Ordered'
    if params["bootstrap_type"] == "Bayesian":
        params["bagging_temperature"] = trial.suggest_float("bagging_temperature", 0, 10)
    elif params["bootstrap_type"] == "Bernoulli":
        params["subsample"] = trial.suggest_float("subsample", 0.5, 1)
    model = catboost.CatBoostClassifier(**params, iterations=10000)
    model.fit(X_train, Y_train, eval_set=(X_test, Y_test), early_stopping_rounds=4000, plot=True, verbose=0)
 
    return model.score(X_test, Y_test)

In [27]:
X_train, X_test, Y_train, Y_test = train_test_split(frst.iloc[:, 2:], frst['CEFR'], test_size=0.25)

In [29]:
study_name = "study_frst" 
storage_name = "sqlite:///{}.db".format(study_name)
study = optuna.create_study(study_name=study_name, storage=storage_name, load_if_exists=True, direction="maximize")
study.optimize(objective, n_trials=10000, timeout=1600)

[I 2022-03-22 02:22:37,674] A new study created in RDB with name: study_frst
Custom logger is already specified. Specify more than one logger at same time is not thread safe.

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 02:25:28,070] Trial 0 finished with value: 0.4097222222222222 and parameters: {'depth': 6, 'learning_rate': 0.10795763372136367, 'l2_leaf_reg': 86494, 'border_count': 964, 'random_strength': 4525, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 68, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'MVS', 'score_function': 'Cosine'}. Best is trial 0 with value: 0.4097222222222222.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 02:27:24,988] Trial 1 finished with value: 0.4027777777777778 and parameters: {'depth': 4, 'learning_rate': 0.04864230963722688, 'l2_leaf_reg': 54422, 'border_count': 3438, 'random_strength': 2848, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 73, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'Bayesian', 'score_function': 'Cosine', 'bagging_temperature': 4.455968060786338}. Best is trial 0 with value: 0.4097222222222222.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 02:30:17,031] Trial 2 finished with value: 0.4097222222222222 and parameters: {'depth': 11, 'learning_rate': 0.09803226670510765, 'l2_leaf_reg': 76403, 'border_count': 9576, 'random_strength': 325, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 28, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 0 with value: 0.4097222222222222.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 02:30:32,632] Trial 3 finished with value: 0.3819444444444444 and parameters: {'depth': 1, 'learning_rate': 0.8096376909271752, 'l2_leaf_reg': 62880, 'border_count': 767, 'random_strength': 2912, 'grow_policy': 'Depthwise', 'leaf_estimation_iterations': 22, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bayesian', 'score_function': 'L2', 'bagging_temperature': 5.508867498527792}. Best is trial 0 with value: 0.4097222222222222.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 02:32:38,593] Trial 4 finished with value: 0.4097222222222222 and parameters: {'depth': 7, 'learning_rate': 0.21235591566905954, 'l2_leaf_reg': 24937, 'border_count': 8062, 'random_strength': 7191, 'grow_policy': 'Depthwise', 'leaf_estimation_iterations': 47, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 0 with value: 0.4097222222222222.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 02:33:22,633] Trial 5 finished with value: 0.3888888888888889 and parameters: {'depth': 4, 'learning_rate': 0.5208536652409143, 'l2_leaf_reg': 42820, 'border_count': 1917, 'random_strength': 6367, 'grow_policy': 'Depthwise', 'leaf_estimation_iterations': 19, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bayesian', 'score_function': 'L2', 'bagging_temperature': 7.4398469259451385}. Best is trial 0 with value: 0.4097222222222222.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 02:33:45,671] Trial 6 finished with value: 0.375 and parameters: {'depth': 2, 'learning_rate': 0.8533431116219206, 'l2_leaf_reg': 63851, 'border_count': 500, 'random_strength': 8826, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 59, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bayesian', 'score_function': 'Cosine', 'bagging_temperature': 3.5923099692643423}. Best is trial 0 with value: 0.4097222222222222.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:14:49,984] Trial 7 finished with value: 0.4166666666666667 and parameters: {'depth': 11, 'learning_rate': 0.8090859481592112, 'l2_leaf_reg': 56648, 'border_count': 3815, 'random_strength': 6033, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 46, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bayesian', 'score_function': 'L2', 'bagging_temperature': 0.995790761595895}. Best is trial 7 with value: 0.4166666666666667.


In [30]:
X_train, X_test, Y_train, Y_test = train_test_split(scnd.iloc[:, 2:], scnd['CEFR'], test_size=0.25)

In [31]:
study_name = "study_scnd" 
storage_name = "sqlite:///{}.db".format(study_name)
study = optuna.create_study(study_name=study_name, storage=storage_name, load_if_exists=True, direction="maximize")
study.optimize(objective, n_trials=10000, timeout=1600)

[I 2022-03-22 03:14:50,236] A new study created in RDB with name: study_scnd


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:19:22,403] Trial 0 finished with value: 0.40939597315436244 and parameters: {'depth': 9, 'learning_rate': 0.961210341245569, 'l2_leaf_reg': 10457, 'border_count': 5225, 'random_strength': 5093, 'grow_policy': 'Depthwise', 'leaf_estimation_iterations': 13, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bernoulli', 'score_function': 'L2', 'subsample': 0.9567207139742355}. Best is trial 0 with value: 0.40939597315436244.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:21:02,132] Trial 1 finished with value: 0.42953020134228187 and parameters: {'depth': 10, 'learning_rate': 0.36720329892806025, 'l2_leaf_reg': 25905, 'border_count': 2944, 'random_strength': 3889, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 91, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 1 with value: 0.42953020134228187.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:22:23,849] Trial 2 finished with value: 0.4228187919463087 and parameters: {'depth': 7, 'learning_rate': 0.8925790506626347, 'l2_leaf_reg': 98729, 'border_count': 3194, 'random_strength': 7355, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 46, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 1 with value: 0.42953020134228187.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:22:59,914] Trial 3 finished with value: 0.4563758389261745 and parameters: {'depth': 2, 'learning_rate': 0.9952788126979148, 'l2_leaf_reg': 51201, 'border_count': 2048, 'random_strength': 1131, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 95, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 3 with value: 0.4563758389261745.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:24:27,202] Trial 4 finished with value: 0.44966442953020136 and parameters: {'depth': 6, 'learning_rate': 0.10451575275549847, 'l2_leaf_reg': 61269, 'border_count': 7331, 'random_strength': 5912, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 73, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bernoulli', 'score_function': 'L2', 'subsample': 0.9421767627988982}. Best is trial 3 with value: 0.4563758389261745.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:33:58,133] Trial 5 finished with value: 0.4228187919463087 and parameters: {'depth': 12, 'learning_rate': 0.2932999115449054, 'l2_leaf_reg': 18151, 'border_count': 845, 'random_strength': 9786, 'grow_policy': 'Depthwise', 'leaf_estimation_iterations': 51, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 3 with value: 0.4563758389261745.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:34:21,579] Trial 6 finished with value: 0.4228187919463087 and parameters: {'depth': 3, 'learning_rate': 0.9936443214748324, 'l2_leaf_reg': 25962, 'border_count': 1501, 'random_strength': 5195, 'grow_policy': 'Lossguide', 'leaf_estimation_iterations': 47, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bernoulli', 'score_function': 'Cosine', 'subsample': 0.5104877556228644}. Best is trial 3 with value: 0.4563758389261745.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:34:41,747] Trial 7 finished with value: 0.42953020134228187 and parameters: {'depth': 3, 'learning_rate': 0.5389250512228111, 'l2_leaf_reg': 29193, 'border_count': 1031, 'random_strength': 9963, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 6, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bernoulli', 'score_function': 'Cosine', 'subsample': 0.7945840594952074}. Best is trial 3 with value: 0.4563758389261745.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:35:03,311] Trial 8 finished with value: 0.4161073825503356 and parameters: {'depth': 2, 'learning_rate': 0.5228655482902081, 'l2_leaf_reg': 68518, 'border_count': 5991, 'random_strength': 8191, 'grow_policy': 'Depthwise', 'leaf_estimation_iterations': 54, 'leaf_estimation_method': 'Gradient', 'bootstrap_type': 'Bernoulli', 'score_function': 'Cosine', 'subsample': 0.6054680763799859}. Best is trial 3 with value: 0.4563758389261745.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:35:56,148] Trial 9 finished with value: 0.4697986577181208 and parameters: {'depth': 3, 'learning_rate': 0.18421905153991897, 'l2_leaf_reg': 60415, 'border_count': 1567, 'random_strength': 1725, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 97, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'Bernoulli', 'score_function': 'L2', 'subsample': 0.9428097321768067}. Best is trial 9 with value: 0.4697986577181208.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:39:07,948] Trial 10 finished with value: 0.436241610738255 and parameters: {'depth': 5, 'learning_rate': 0.02319071163548672, 'l2_leaf_reg': 85337, 'border_count': 9973, 'random_strength': 905, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 80, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'Bayesian', 'score_function': 'Cosine', 'bagging_temperature': 0.23757440129614604}. Best is trial 9 with value: 0.4697986577181208.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:39:43,580] Trial 11 finished with value: 0.4228187919463087 and parameters: {'depth': 1, 'learning_rate': 0.7413655867668221, 'l2_leaf_reg': 41829, 'border_count': 3034, 'random_strength': 246, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 100, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'Bayesian', 'score_function': 'L2', 'bagging_temperature': 9.652854826908097}. Best is trial 9 with value: 0.4697986577181208.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:40:25,489] Trial 12 finished with value: 0.47651006711409394 and parameters: {'depth': 4, 'learning_rate': 0.7147999829852767, 'l2_leaf_reg': 50276, 'border_count': 280, 'random_strength': 2349, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 73, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 12 with value: 0.47651006711409394.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:40:57,363] Trial 13 finished with value: 0.4563758389261745 and parameters: {'depth': 5, 'learning_rate': 0.664206176106063, 'l2_leaf_reg': 72577, 'border_count': 23, 'random_strength': 2717, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 69, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'MVS', 'score_function': 'L2'}. Best is trial 12 with value: 0.47651006711409394.


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

[I 2022-03-22 03:41:59,005] Trial 14 finished with value: 0.3959731543624161 and parameters: {'depth': 4, 'learning_rate': 0.23867268115104223, 'l2_leaf_reg': 48731, 'border_count': 4133, 'random_strength': 2564, 'grow_policy': 'SymmetricTree', 'leaf_estimation_iterations': 86, 'leaf_estimation_method': 'Newton', 'bootstrap_type': 'Bayesian', 'score_function': 'L2', 'bagging_temperature': 4.530508542535601}. Best is trial 12 with value: 0.47651006711409394.
